### PR TITLE
Try to generate a random number when minting an nft

### LINKED ISSUE #105: Review random number keeper function: Auto-generate random number?  Not only owner?

### PR DESCRIPTION
**Pull request by slorenzo**
Closes #105

### CODE REVIEW

**Review by ajsantander — APPROVED** [review]

File: packages/hardhat/contracts/Ethernauts.sol [hunk 1]
```diff
@@ -226,11 +211,30 @@ contract Ethernauts is ERC721Enumerable, Ownable {
...
+        uint maxTokenIdInBatch = batchSize * (randomNumberIdx + 1) - 1;
+        if (totalSupply() < maxTokenIdInBatch) {
+            return;
+        }
+
+        // solhint-disable not-rely-on-time
+        uint randomNumber = uint256(
+            keccak256(abi.encodePacked(msg.sender, block.difficulty, block.timestamp, _randomNumbers.length))
+        );
+        // solhint-enable not-rely-on-time
```
Comment (mjlescano) [inline packages/hardhat/contracts/Ethernauts.sol:236]: Little comment nit:
```suggestion
        // solhint-disable-next-line not-rely-on-time
        uint randomNumber = uint256(
            keccak256(abi.encodePacked(msg.sender, block.difficulty, block.timestamp, _randomNumbers.length))
        );
```

File: packages/hardhat/contracts/Ethernauts.sol [hunk 2]
```diff
@@ -226,11 +211,30 @@ contract Ethernauts is ERC721Enumerable, Ownable {
...
+        uint randomNumberIdx = _randomNumbers.length;
+
+        uint maxTokenIdInBatch = batchSize * (randomNumberIdx + 1) - 1;
+        if (totalSupply() < maxTokenIdInBatch) {
+            return;
+        }
+
+        // solhint-disable not-rely-on-time
+        uint randomNumber = uint256(
+            keccak256(abi.encodePacked(msg.sender, block.difficulty, block.timestamp, _randomNumbers.length))
```
Comment (mariobyn) [inline packages/hardhat/contracts/Ethernauts.sol:234]: it's more gas efficient to use randomNumberIdx than call again _randomNumbers.length

Comment (GalloDaSballo) [reply packages/hardhat/contracts/Ethernauts.sol:234]: Agree that the cached value saves a read from storage

Comment (GalloDaSballo) [reply packages/hardhat/contracts/Ethernauts.sol:234]: Notice that all the variables are known at time of minting, the number is not random

Comment (ajsantander) [reply packages/hardhat/contracts/Ethernauts.sol:234]:
```suggestion
            keccak256(abi.encodePacked(msg.sender, block.difficulty, block.timestamp, randomNumberIdx))
```

**Pull request merged**